src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -160,22 +162,27 @@ def _get_source_file(assembly_dir: str, assembly_file: str) -> str:
     return src_file
 
 
-def _upload_assembly_to_workspace(
-    conf: Conf,
-    workspace_id: int,
-    file_path: str,
-    assembly_name: str,
-) -> str:
-    """Upload an assembly file to workspace."""
+def _upload_assemblies_to_workspace(
+        conf: Conf,
+        workspace_id: int,
+        file_paths: tuple[str, ...],
+        assembly_names: tuple[str, ...],
+) -> tuple[str, ...]:
+    """
+    Upload assembly files to the target workspace in batch. The bulk method fails
+    and an error will be thrown if any of the assembly files in batch fails to upload.
+    """
+    inputs = [{"file": f, "assembly_name": n}
+              for f, n in zip(file_paths, assembly_names)]
+
     success, attempts, max_attempts = False, 0, 3
     while attempts < max_attempts and not success:
```

Comment:
This is tricky now since on failure some of the assemblies might have succeeded. I think that what needs to be done is on a failure, look up the assemblies in the batch in the workspace by object name (which I assume is derived from the file name) and remove any files that were successfully uploaded from the list.

We'd be assuming we have control over the workspace and once an assembly from a particular file name was uploaded it's done forever, but I *think* that's ok in this context. We should document that though.

Reply:
> This is tricky now since on failure some of the assemblies might have succeeded.

I tested on this Tuesday already. 

There is no partial upload. A single assembly failure would cause batch failure. I guess no extra lookup in the workspace is needed.

Reply:
As the author of the workspace I can assure you that partial saves are possible. The workspace does what it can to prevent it, but if the server or the database goes down mid save, for example, a partial save could occur

Reply:
:+1: